knodle/trainer/crossweigh_weighing/data_splitting_by_rules.py

```diff
@@ -0,0 +1,289 @@
+
+    else:
+        samples_id_rules_dict = dict((i, set()) for i in range(0, rule_matches_z.shape[0]))
+        for row_id, row in enumerate(rule_matches_z):
+            samples_id_rules_dict[row_id] = np.where(row == 1)[0].tolist()
+
+    for sample_id, rules in samples_id_rules_dict.items():
+        signature = "_".join(map(str, sorted(list(rules))))
+
+        if signature in signature2id:
+            signature_id = signature2id.get(signature)
+        else:
+            signature_id = len(signature2id)
+            signature2id[signature] = signature_id
+
+        if signature_id in signature_id2samples:
+            signature_id2samples[signature_id].append(sample_id)
+        else:
+            signature_id2samples[signature_id] = [sample_id]
+
+    return signature2id, signature_id2samples
+
+
+def compose_train_n_test_datasets(
+        data_features: np.ndarray, rule_id2samples_ids: Dict, labels: np.ndarray, folds: int,
+        partitions: int, other_class_id: int = None
+) -> (List, List):
+    """
+    This function creates train and test datasets for k-folds cross-validation.
+
+    :param data_features: encoded data samples (num_samples x num_features):
+    :param rule_id2samples_ids: {rule_id: [sample_id1, sample_id2, ...]}. Rule in this context means anything basing on
+    which splitting is performed (matched rule, sample signature, ...).
+    :param labels: array of labels
+    :param folds: number of folds the data instances are to be slitted into in each partition
+    :param partitions: number of partitions that are to be performed
+    :param other_class_id: if you don't want to include the negative samples (the ones that belong to the other class)
+     to the test set, but only to the training set, you can pass the id of other class should be
+
+    :return: list of train sets and list of corresponding test (hold-out) sets
+    """
+
+    # calculate ids of all samples that belong to the 'other_class'
+    other_sample_ids = np.where(labels[:, other_class_id] == 1)[0].tolist() if other_class_id else None
+    # make a list of rule ids to shuffle them later
+    rule_ids = [rule_id for rule_id in range(0, len(rule_id2samples_ids))]
+
+    train_datasets, test_datasets = [], []
+    for partition in range(partitions):
+        logger.info(f"Partition {partition + 1}/{partitions}:")
+        random.shuffle(rule_ids)  # shuffle anew for each splitting
+        for fold in range(folds):
+            train_dataset, test_dataset = get_train_test_datasets_by_rule_indices(
+                data_features, rule_ids, rule_id2samples_ids, labels, fold, folds, other_sample_ids
+            )
+            train_datasets.append(train_dataset)
+            test_datasets.append(test_dataset)
+
+    return train_datasets, test_datasets
+
+
+def get_train_test_datasets_by_rule_indices(
+        data_features: np.ndarray, rules_ids: List[int], rule_id2samples_ids: Dict, labels: np.ndarray, fold: int,
```

Comment:
Not neccessarily but to avoid confusion with "folds"
```suggestion
        data_features: np.ndarray, rules_ids: List[int], rule_id2samples_ids: Dict, labels: np.ndarray, fold_id: int,
```